src/Compilers/Core/AnalyzerDriver/AnalyzerManager.cs

```diff
@@ -0,0 +1,138 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.CodeAnalysis.Diagnostics
+{
+    /// <summary>
+    /// Manages properties of analyzers (such as registered actions, supported diagnostics) for analyzer host's lifetime.
+    /// 
+    /// It ensures the following for the lifetime of analyzer host:
+    /// 1) <see cref="DiagnosticAnalyzer.Initialize(AnalysisContext)"/> is invoked only once per-analyzer.
+    /// 2) <see cref="DiagnosticAnalyzer.SupportedDiagnostics"/> is invoked only once per-analyzer.
+    /// 3) <see cref="CompilationStartAnalyzerAction"/> registered during Initialize are invoked only once per-analyzer per-compilation.
+    /// </summary>
+    internal class AnalyzerManager
+    {
+        public static readonly AnalyzerManager Default = new AnalyzerManager();
+
+        // This map stores the tasks to compute HostSessionStartAnalysisScope for session wide analyzer actions, i.e. AnalyzerActions registered by analyzer's Initialize method.
+        // These are run only once per every analyzer.
+        private ImmutableDictionary<DiagnosticAnalyzer, Task<HostSessionStartAnalysisScope>> _sessionScopeMap =
+            ImmutableDictionary<DiagnosticAnalyzer, Task<HostSessionStartAnalysisScope>>.Empty;
+
+        // This map stores the tasks to compute HostCompilationStartAnalysisScope for per-compilation analyzer actions, i.e. AnalyzerActions registered by analyzer's CompilationStartActions.
+        // Compilation start actions will get executed once per-each compilation as user might want to return different set of custom actions for each compilation.
+        private readonly ConditionalWeakTable<Compilation, ConcurrentDictionary<DiagnosticAnalyzer, Task<HostCompilationStartAnalysisScope>>> _compilationScopeMap =
+            new ConditionalWeakTable<Compilation, ConcurrentDictionary<DiagnosticAnalyzer, Task<HostCompilationStartAnalysisScope>>>();
```

Comment:
set concurrency level to 2. otherwise, concurrent dictionary allocates too much stuff to reduce contention based on number of threads the dictionary is used. problem is since we use threadpool tasks, it runs on many many different threads and it confuse the dictionary. and 2 concurrent threads changing the dictionary is most likely enough for us.